feat: Add fields to User table and registration

bot190 opened the pull request:
Adds the following fields to the User Table and User Registration page.
- Preferred Name
- Phone Number
- Address
- Email
- T-Shirt Size 

Also added a Subteam selection to the registration page.

bot190 committed:
feat: Add fields to User table and registration

Adds the following fields to the User Table and User Registration page.
- Preferred Name
- Phone Number
- Address
- Email
- T-Shirt Size 

Also added a Subteam selection to the registration page.

msoucy commented on signinapp/model.py:
```diff
@@ -51,14 +52,33 @@ class StampEvent():
     __tablename__ = "users"
     id = db.Column(db.Integer, primary_key=True)
     name = db.Column(db.String, nullable=False)
+    preferred_name = db.Column(db.String, nullable=True)
```
Why is this `nullable=True` if the form has it as required?

bot190 replied:
I've removed the `nullable=True` since it's the default (I think). I've also made this field optional as we will use the name instead if nothing is provided.

msoucy commented on signinapp/model.py:
```diff
@@ -381,6 +413,10 @@ def from_name(cls, name) -> Subteam:
         return cls.query.filter_by(name=name).one_or_none()
 
+    @classmethod
+    def get_subteam(cls) -> List[str]:
```
This isn't getting one subteam, this is getting all subteams. `get_subteams` instead (plural)

But is it even used?

bot190 replied:
Updating this to match what you're doing in `admin.users.new_user` and `admin.users.edit_user`, and replacing those uses with calls to this function to ensure we're consistent.

msoucy reviewed (changes requested):
If this gets run, will it break all of the existing databases?

bot190 committed:
fix: Normalize getting form subteam listing

Use the same query / comprehension to create the list of subteams for the subteam select field throughout app.

bot190 committed:
feat: Add support for new fields to admin paths

Adds support for the new fields to the "new_user" and "edit_user" paths available to mentors.

bot190 committed:
test: Add more details for Matt's test user

Adds more details for Matt's user to ensure the additional fields show up correctly.

msoucy commented on signinapp/admin/users.py:
```diff
@@ -3,19 +3,27 @@
+
+    phone_number = TelField("Phone Number")
+    email = EmailField("Email Address")
+    address = StringField("Street Address")
```
*Technically* other teams (or even us in weird cases) might want to have the address include the town, but since this is just a string it's fine as is.

bot190 replied:
We could change this to just Address? The street is hopefully all we care about, and I don't want/need everyone spending a bunch of time entering their full address. I'll leave as-is for now, we can change the form in the future.

msoucy commented on signinapp/admin/users.py:
```diff
@@ -59,10 +66,19 @@ def new_user():
             user.subteam = Subteam.query.get(form.subteam.data)
+
+        # If no preferred name was provided default to their name
+        if not form.preferred_name.data:
```
We talked about it offline, but for posterity - `preferred_name` should be nullable, and an accessor (`@hybrid_property`) should be made `display_name` that uses preferred if it exists or falls back to `name`

msoucy commented on signinapp/admin/users.py:
```diff
@@ -77,15 +93,14 @@ def new_user():
 @admin_required
 def edit_user():
-    user : User = User.query.get(request.args["user_id"])
+    user: User = User.query.get(request.args["user_id"])
```
Please use `db.session.get(User, request.args["user_id"])` for this - see [here](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-get-method-moves-to-session) where the query object is deprecated

msoucy commented on signinapp/admin/users.py:
```diff
@@ -96,6 +111,17 @@ def edit_user():
         user.active = form.active.data
+
+        # If no preferred name was provided default to their name
+        if form.preferred_name.data:
```
"Display name" comment affects here too

msoucy commented on signinapp/auth.py:
```diff
@@ -50,10 +60,18 @@ class ChangePasswordForm(FlaskForm):
+        address = form.address.data
+        tshirt_size = ShirtSizes[form.tshirt_size.data]
+
+        subteam = Subteam.query.get(form.subteam.data)
```
`db.session.get(Subteam, form.subteam.data)`

msoucy commented on signinapp/auth.py:
```diff
@@ -50,10 +60,18 @@ class ChangePasswordForm(FlaskForm):
         # code to validate and add user to database goes here
         name = form.name.data
         password = form.password.data
+        preferred_name = form.preferred_name.data
```
Thinking about it, can we move this block to after the check for a user already existing?

msoucy commented on signinapp/model.py:
```diff
@@ -125,14 +145,26 @@ def human_readable(self) -> str:
                     approved=approved)
 
-    @staticmethod
+    @ staticmethod
```
Stray space

msoucy commented on signinapp/model.py:
```diff
@@ -125,14 +145,26 @@ def human_readable(self) -> str:
 
     @staticmethod
-    def make(name: str, password: str, role: Role,
+    def make(name: str, password: str, role: Role, preferred_name: str = None,
```
This is getting excessive. Can we use something like this?

```py
@functools.wraps(User.__init__)
@staticmethod
def make(name: str, password: str, role: Role, **kwargs):
    return User(name=name, password=generate_password_hash(password),
                code=mk_hash(name), role_id=Role, **kwargs)
    
```

bot190 committed:
chore: Clean-up code per review comments

Allow preferred_name to be none, and add display_name as a hybrid property that will default to the full name if the preferred_name is Null.

Clean-up some SQL queries to use the newer format.

msoucy commented on signinapp/model.py:
```diff
@@ -122,15 +142,20 @@ def can_view(self, user: User):
+
+    @hybrid_property
+    def display_name(self) -> str:
+        return self.preferred_name if self.preferred_name else self.name
```
`return self.preferred_name or self.name`
